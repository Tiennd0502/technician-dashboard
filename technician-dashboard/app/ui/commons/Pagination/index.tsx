'use client';

import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
import { Flex, Button } from '@chakra-ui/react';
import React, { useCallback, useMemo } from 'react';

interface PaginationProps {
  total: number;
  page: number;
  onChange: (value: number) => void;
}

const Pagination = ({ total, page, onChange }: PaginationProps) => {
  const handleClickNext = useCallback(() => {
    onChange(page + 1);
  }, [onChange, page]);

  const handleClickPrev = useCallback(() => {
    onChange(page - 1);
  }, [onChange, page]);

  const handleClickFirst = useCallback(() => {
    onChange(1);
  }, [onChange]);

  const handleClickLast = useCallback(() => {
    onChange(total);
  }, [total, onChange]);

  const pages = useMemo(() => {
    if (page <= 4 || total < 8) {
      if (total >= 8) {
        return [2, 3, 4, 5, '...'];
      } else {
        const result = Array.from({ length: total }, (_, i) => i + 1);
        return result.slice(1, result.length - 1);
      }
    } else if (total - page <= 3) {
      return ['...', total - 4, total - 3, total - 2, total - 1];
    } else {
      return ['...', page - 1, page, page + 1, '...'];
    }
  }, [total, page]);

  return (
    <Flex columnGap='2' justifyContent='flex-end' w='full' role='list'>
      <Button variant='outline' isDisabled={page <= 1} w='42px' onClick={handleClickPrev}>
        <ChevronLeftIcon fontSize={20} />
      </Button>
      <Button w='42px' variant={page === 1 ? 'solid' : 'outline'} onClick={handleClickFirst}>
        1
      </Button>
      {pages.map((item, index) => {
        const handleChangePage = () => {
          if (typeof item === 'number' && item !== page) {
            onChange(item);
          }
        };

        return (
          <Button
            w='42px'
            variant={page === +item ? 'solid' : 'outline'}
            key={`${item}${index}`}
            onClick={handleChangePage}
            isDisabled={typeof item === 'string'}
          >
            {item}
          </Button>
        );
      })}

      {total >= 2 && (
        <Button w='42px' variant={page === total ? 'solid' : 'outline'} onClick={handleClickLast}>
          {total}
        </Button>
      )}
      <Button variant='outline' isDisabled={page === total} w='42px' onClick={handleClickNext}>
        <ChevronRightIcon fontSize={20} />
      </Button>
    </Flex>
  );
};

export default Pagination;
