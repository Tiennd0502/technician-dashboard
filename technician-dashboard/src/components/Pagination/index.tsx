'use client';

import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
import { Flex, Button } from '@chakra-ui/react';
import React, { useCallback, useMemo, useState } from 'react';

interface PaginationProps {
  total: number;
  page: number;
  onChange: (value: number) => void;
}

const Pagination = ({ total, page, onChange }: PaginationProps) => {
  const [currentPage, setCurrentPage] = useState(page);

  const handleClickNext = useCallback(() => {
    setCurrentPage((prev) => ++prev);
  }, []);

  const handleClickPrev = useCallback(() => {
    setCurrentPage((prev) => --prev);
  }, []);

  const handleClickFirst = useCallback(() => {
    setCurrentPage(1);
  }, []);

  const handleClickLast = useCallback(() => {
    setCurrentPage(total);
  }, [total]);

  const pages = useMemo(() => {
    if (currentPage <= 4) {
      return [2, 3, 4, 5, '...'];
    } else if (total - currentPage <= 2) {
      return ['...', total - 4, total - 3, total - 2, total - 1];
    } else {
      return ['...', currentPage - 1, currentPage, currentPage + 1, '...'];
    }
  }, [total, currentPage]);

  return (
    <Flex columnGap='2' justifyContent='flex-end' w='full' role='list'>
      <Button variant='outline' isDisabled={currentPage <= 1} w='42px' onClick={handleClickPrev}>
        <ChevronLeftIcon fontSize={20} />
      </Button>
      <Button w='42px' variant={currentPage === 1 ? 'solid' : 'outline'} onClick={handleClickFirst}>
        1
      </Button>
      {pages.map((item, index) => {
        const handleChangePage = () => {
          if (typeof item === 'number' && item !== currentPage) {
            setCurrentPage(item);
            onChange(item);
          }
        };

        return (
          <Button
            w='42px'
            variant={currentPage === +item ? 'solid' : 'outline'}
            key={`${item}${index}`}
            onClick={handleChangePage}
            isDisabled={typeof item === 'string'}
          >
            {item}
          </Button>
        );
      })}
      <Button
        w='42px'
        variant={currentPage === total ? 'solid' : 'outline'}
        onClick={handleClickLast}
      >
        {total}
      </Button>
      <Button
        variant='outline'
        isDisabled={currentPage === total}
        w='42px'
        onClick={handleClickNext}
      >
        <ChevronRightIcon fontSize={20} />
      </Button>
    </Flex>
  );
};

export default Pagination;
