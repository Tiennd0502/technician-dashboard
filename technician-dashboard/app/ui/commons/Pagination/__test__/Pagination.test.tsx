import { render } from '@testing-library/react';

import Pagination from '..';

describe('Pagination test cases', () => {
  const props = {
    page: 1,
    total: 10,
    onChange: jest.fn(),
  };

  test('should render correctly', () => {
    const { container } = render(<Pagination {...props} />);
    expect(container).toMatchSnapshot();
  });

  test('Should render enough elements', () => {
    const { getByTestId } = render(<Pagination {...props} />);
    const pageList = getByTestId('list-page');

    expect(pageList.childNodes).toHaveLength(9);
  });
});
