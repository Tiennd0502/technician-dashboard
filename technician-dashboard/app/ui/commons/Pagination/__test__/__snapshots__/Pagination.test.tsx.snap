// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test cases should render correctly 1`] = `
<div>
  <div
    class="css-16kx4g3"
    role="list"
  >
    <button
      class="chakra-button css-17rlipp"
      disabled=""
      type="button"
    >
      <svg
        class="chakra-icon css-14bgo1y"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          fill="currentColor"
        />
      </svg>
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      1
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      2
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      3
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      4
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      5
    </button>
    <button
      class="chakra-button css-17rlipp"
      disabled=""
      type="button"
    >
      ...
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      10
    </button>
    <button
      class="chakra-button css-17rlipp"
      type="button"
    >
      <svg
        class="chakra-icon css-14bgo1y"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
